File: vbnet/metafiles/metafiles.vb
' metafiles -- VB.NET port of examples\Vb6\metafiles
Imports System
Imports LumasPdfSdk

Module modMetafiles
    Private errDel As TErrorProc
    Private Const CLR_RED As UInteger = 255UI
    Private Const MARGIN As Double = 10.0

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Public Function ErrProc(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return 0
    End Function

    Private Sub PlaceEMFCentered(ByVal pdf As IntPtr, ByVal MFile As String, ByVal Width As Double, ByVal Height As Double)
        Dim x, y, w, h, sx As Double
        Dim r As TRectL
        LumasPdf.pdfGetLogMetafileSizeW(pdf, MFile, r)
        w = r.Right - r.Left
        h = r.Bottom - r.Top
        Width = Width - 2.0 * MARGIN
        Height = Height - 2.0 * MARGIN
        sx = Width / w
        If (h * sx <= Height) Then
            x = MARGIN
            h = h * sx
            y = (Height - h) / 2.0
            LumasPdf.pdfInsertMetafileW(pdf, MFile, x, y, Width, 0.0)
            LumasPdf.pdfSetStrokeColor(pdf, CLR_RED)
            LumasPdf.pdfRectangle(pdf, x, y, Width, h, TPathFillMode.fmStroke)
        Else
            sx = Height / h
            w = w * sx
            x = (Width - w) / 2.0
            y = MARGIN
            LumasPdf.pdfInsertMetafileW(pdf, MFile, x, y, 0.0, Height)
            LumasPdf.pdfSetStrokeColor(pdf, CLR_RED)
            LumasPdf.pdfRectangle(pdf, x, y, w, Height, TPathFillMode.fmStroke)
        End If
    End Sub

    Sub Main()
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        errDel = New TErrorProc(AddressOf ErrProc)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)
        If LumasPdf.pdfCreateNewPDFW(pdf, "") = 0 Then
            LumasPdf.pdfDeletePDF(pdf)
            Return
        End If

        LumasPdf.pdfSetPageCoords(pdf, TPageCoord.pcTopDown)

        LumasPdf.pdfAppend(pdf)
        LumasPdf.pdfSetOrientationEx(pdf, 90)
        PlaceEMFCentered(pdf, AppPath() & "\coords.emf", LumasPdf.pdfGetPageWidth(pdf), LumasPdf.pdfGetPageHeight(pdf))
        LumasPdf.pdfEndPage(pdf)

        LumasPdf.pdfAppend(pdf)
        LumasPdf.pdfSetOrientationEx(pdf, 90)
        PlaceEMFCentered(pdf, AppPath() & "\fulltest.emf", LumasPdf.pdfGetPageWidth(pdf), LumasPdf.pdfGetPageHeight(pdf))
        LumasPdf.pdfEndPage(pdf)

        LumasPdf.pdfAppend(pdf)
        LumasPdf.pdfSetOrientationEx(pdf, 90)
        PlaceEMFCentered(pdf, AppPath() & "\gdi.emf", LumasPdf.pdfGetPageWidth(pdf), LumasPdf.pdfGetPageHeight(pdf))
        LumasPdf.pdfEndPage(pdf)

        Dim outFile As String = AppPath() & "\out.pdf"
        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            If LumasPdf.pdfOpenOutputFileW(pdf, outFile) = 0 Then
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
        End If
        If LumasPdf.pdfCloseFile(pdf) <> 0 Then
            Console.WriteLine("PDF file """ & outFile & """ successfully created!")
        End If
        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module

File: cpp-linux/metafiles/metafiles.cpp
// metafiles -- C++ port of examples\Vb6\metafiles\metafiles.bas
// Places three EMF metafiles, each centered and scaled to a landscape page,
// with a red frame around them.
#include <lumaspdf.h>
#include "repo_root.h"
#include <cstdio>
#include <string>

static const UI32 CLR_RED = 255;
static const double MARGIN = 10.0;
static const char* DIR = LUMAS_REPO_ROOT "/";

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static void PlaceEMFCentered(PPDF pdf, const char* mFile, double width, double height){
    TRectL r{};
    pdfGetLogMetafileSizeA(pdf, mFile, &r);
    double w = r.Right - r.Left;
    double h = r.Bottom - r.Top;
    width  -= 2.0 * MARGIN;
    height -= 2.0 * MARGIN;
    double sx = width / w;
    if(h * sx <= height){
        double x = MARGIN;
        h = h * sx;
        double y = (height - h) / 2.0;
        pdfInsertMetafileA(pdf, mFile, x, y, width, 0.0);
        pdfSetStrokeColor(pdf, CLR_RED);
        pdfRectangle(pdf, x, y, width, h, fmStroke);
    } else {
        sx = height / h;
        w = w * sx;
        double x = (width - w) / 2.0;
        double y = MARGIN;
        pdfInsertMetafileA(pdf, mFile, x, y, 0.0, height);
        pdfSetStrokeColor(pdf, CLR_RED);
        pdfRectangle(pdf, x, y, w, height, fmStroke);
    }
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, ErrProc);
    if(pdfCreateNewPDFA(pdf, "") == 0){ pdfDeletePDF(pdf); return 1; }

    pdfSetPageCoords(pdf, pcTopDown);

    const char* names[3] = { "sample_vector.emf", "sample_vector.emf", "sample_vector.emf" };
    for(int i = 0; i < 3; ++i){
        std::string mf = std::string(DIR) + names[i];
        pdfAppend(pdf);
        pdfSetOrientationEx(pdf, 90);
        PlaceEMFCentered(pdf, mf.c_str(), pdfGetPageWidth(pdf), pdfGetPageHeight(pdf));
        pdfEndPage(pdf);
    }

    std::string outFile = exeDir(argv[0]) + "/out.pdf";
    if(pdfHaveOpenDoc(pdf) != 0){
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){ pdfDeletePDF(pdf); return 1; }
    }
    if(pdfCloseFile(pdf) != 0) printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
    pdfDeletePDF(pdf);
    return 0;
}
